feat: Add AIStupidLevel provider integration

StudioPlatforms opened the pull request:
## Summary

This PR adds AIStupidLevel as a new provider to Kilo Code, enabling users to leverage an intelligent AI router that automatically selects the best-performing model based on real-time benchmarks.

## What is AIStupidLevel?

AIStupidLevel is an intelligent AI router that continuously benchmarks 25+ AI models across multiple providers (OpenAI, Anthropic, Google, xAI, DeepSeek, and more) and automatically routes requests to the best-performing model based on real-time performance data.

## Changes Made

### 1. Provider Type Definitions (`packages/types/src/provider-settings.ts`)
- Added `aistupidlevel` to the `dynamicProviders` array
- Created `aiStupidLevelSchema` with API key and model ID configuration
- Added AIStupidLevel to the discriminated union of provider settings
- Added `aiStupidLevelModelId` to `modelIdKeys`
- Added AIStupidLevel to `modelIdKeysByProvider` mapping
- Added AIStupidLevel to `MODELS_BY_PROVIDER` with label "AIStupidLevel"
- Added AIStupidLevel schema to `providerSettingsSchema`

### 2. API Handler (`src/shared/api.ts`)
- Added AIStupidLevel to `dynamicProviderExtras` with optional API key parameter

### 3. Documentation (`apps/kilocode-docs/docs/providers/aistupidlevel.md`)
- Created comprehensive documentation explaining:
  - What AIStupidLevel is and how it works
  - How to get an API key
  - Available routing strategies (6 different auto models)
  - Configuration steps in Kilo Code
  - Key features and benefits
  - Response headers for transparency
  - Pricing information
  - Tips and best practices

## Available Routing Strategies

AIStupidLevel offers 6 intelligent routing strategies:

- `auto` - Best overall performance across all metrics
- `auto-coding` - Optimized for code generation and quality
- `auto-reasoning` - Best for complex reasoning and problem-solving
- `auto-creative` - Optimized for creative writing quality
- `auto-cheapest` - Most cost-effective option
- `auto-fastest` - Fastest response time

## Key Features

- **Real-time performance benchmarking** with 7-axis scoring methodology
- **Statistical degradation detection** to avoid poorly performing models
- **Cost optimization** with automatic provider switching
- **Transparent routing** with detailed response headers
- **Enterprise SLA** with 99.9% uptime guarantee
- **Multi-provider support** (OpenAI, Anthropic, Google, xAI, DeepSeek, etc.)

## Testing

The integration follows the same pattern as other dynamic providers in Kilo Code (OpenRouter, Vercel AI Gateway, etc.) and should work seamlessly with the existing infrastructure.

## Documentation

Full documentation has been added at `apps/kilocode-docs/docs/providers/aistupidlevel.md` following the same format as other provider documentation.

## Links

- **Website:** https://aistupidlevel.info
- **Router Dashboard:** https://aistupidlevel.info/router
- **Community:** https://www.reddit.com/r/AIStupidlevel
- **Twitter/X:** https://x.com/AIStupidlevel

StudioPlatforms committed:
feat: Add AIStupidLevel provider integration

- Added AIStupidLevel as a dynamic provider in provider-settings.ts
- Created schema and configuration for AIStupidLevel
- Added AIStupidLevel to MODELS_BY_PROVIDER and modelIdKeysByProvider
- Updated api.ts to include AIStupidLevel in dynamicProviderExtras
- Created comprehensive documentation for AIStupidLevel provider

AIStupidLevel is an intelligent AI router that automatically selects
the best-performing model based on real-time benchmarks across 25+ AI
models from multiple providers (OpenAI, Anthropic, Google, xAI, etc.).

Features:
- 6 routing strategies (auto, auto-coding, auto-reasoning, auto-creative, auto-cheapest, auto-fastest)
- Real-time performance benchmarking with 7-axis scoring
- Statistical degradation detection
- Cost optimization with automatic provider switching
- Transparent routing decisions

StudioPlatforms committed:
fix: Resolve merge conflict - add local providers from main

- Added lmstudio and ollama to MODELS_BY_PROVIDER
- Maintains aistupidlevel provider integration
- Resolves conflict with main branch changes

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

mcowger reviewed (changes requested):
This pull request appears to be incomplete?   

There's docs and a couple added settings, but no UI to configure, no tests, no provider implementation, etc.

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
feat: Complete AIStupidLevel provider implementation

This commit completes the AIStupidLevel provider integration with all required components:

Backend Implementation:
- Add AIStupidLevelHandler extending RouterProvider
- Implement model fetcher with fallback to default routing strategies
- Integrate into buildApiHandler and model cache system
- Support for streaming and single completion modes

Type Definitions:
- Add provider settings schema and validation
- Define default model constants (auto-coding as default)
- Export types from providers index

UI Components:
- Create AIStupidLevel settings component with API key input
- Add model picker for routing strategy selection
- Integrate into ApiOptions with proper routing
- Add to provider constants list

Translations:
- Add English translation keys for API key labels

Features:
- Smart routing with multiple strategies (auto, auto-coding, auto-reasoning, etc.)
- OpenAI-compatible API integration
- Fallback models when API fetch fails
- Full streaming support
- Proper error handling

This addresses the maintainer feedback about the incomplete PR by providing:
✓ Complete provider implementation
✓ UI configuration components
✓ Proper integration throughout codebase
✓ Type-safe implementation

StudioPlatforms commented:
Hi @mcowger,

Sorry about the incomplete initial submission! I've now pushed a complete implementation that addresses all your feedback:

## What's Been Added:

✅ **Complete Provider Implementation**
- AIStupidLevelHandler extending RouterProvider
- Full OpenAI-compatible API integration at https://api.aistupidlevel.info/v1
- Support for streaming and single completion modes

✅ **Model Fetcher**
- Fetches available routing strategies from the API
- Fallback to default models (auto, auto-coding, auto-reasoning, etc.)
- Integrated into the model cache system

✅ **UI Configuration**
- Complete settings component with API key input
- Model picker for selecting routing strategies
- Proper integration into ApiOptions
- Added to provider constants list

✅ **Type Definitions**
- Provider settings schema with validation
- Default model constants
- Proper exports throughout the type system

✅ **Translations**
- English translation keys for UI labels

The implementation now follows the same patterns as other providers in the codebase (like VercelAiGateway, DeepInfra, etc.) and should be ready for review. Let me know if you need any changes!

Thanks for your patience!

mcowger commented on apps/kilocode-docs/docs/providers/aistupidlevel.md:
```diff
@@ -0,0 +1,102 @@
+- **Provider Diversity:** Access models from OpenAI, Anthropic, Google, xAI, DeepSeek, and more through one API
+- **Transparent Routing:** Response headers show which model was selected and why
+- **Performance Tracking:** Dashboard shows your usage, cost savings, and routing decisions
+- **Enterprise SLA:** 99.9% uptime guarantee with multi-region deployment
```
Rather than including the marketing copy here, could you include a link to your own pages?

mcowger commented on apps/kilocode-docs/docs/providers/aistupidlevel.md:
```diff
@@ -0,0 +1,102 @@
+
+## Tips and Notes
+
+- **Provider Keys Required:** You must add your own provider API keys (OpenAI, Anthropic, etc.) to your AIStupidLevel dashboard before using the router
```
This should go into the 'Configuration' section above given that its a required activity.

mcowger commented on packages/types/src/providers/aistupidlevel.ts:
```diff
@@ -0,0 +1,16 @@
+	supportsImages: true,
+	supportsComputerUse: false,
+	supportsPromptCache: false,
+	inputPrice: 0,
```
Can we set correct default prices here?

mcowger commented on src/api/providers/fetchers/aistupidlevel.ts:
```diff
@@ -0,0 +1,151 @@
+ * getAIStupidLevelModels
+ */
+
+export async function getAIStupidLevelModels(apiKey?: string): Promise<Record<string, ModelInfo>> {
```
Given that this is required to get anything useful, this should be a required parameter

mcowger commented on src/api/providers/fetchers/aistupidlevel.ts:
```diff
@@ -0,0 +1,151 @@
+ * parseAIStupidLevelModel
+ */
+
+export const parseAIStupidLevelModel = (model: AIStupidLevelModel): ModelInfo => {
```
Good type safety!

mcowger commented on src/api/providers/aistupidlevel.ts:
```diff
@@ -0,0 +1,98 @@
+		})
+	}
+
+	override async *createMessage(
```
There's no non-streaming path implemented here.   But your implementation of completePrompt() below certainly implies your service supports it.

Is it intentional to only support streaming for agentic activity?

mcowger reviewed (changes requested):
This is overall a really solid first PR.  Nicely done.

A couple overall things beyond the small comments made on specific lines:

1. Need to add //kilocode_change markers for changes to existing files (net new files dont need them).  These are used to make merging from Roo/Cline easier.
  a. Block markers: "// kilocode_change start" ... "// kilocode_change end" — used to mark a region of added/modified code.
  b. Inline markers: "// kilocode_change", or comment suffixes like "// kilocode_change: explanation" — used to flag single-line changes or explain why a change was made.

2. There are no tests provided.  See some examples in the PR for Synthetic: https://github.com/Kilo-Org/kilocode/pull/2567/files

StudioPlatforms committed:
fix: Add kilocode_change markers and comprehensive tests

Addresses maintainer feedback:

1. Added // kilocode_change markers to all modified files:
   - src/api/index.ts
   - src/api/providers/index.ts
   - src/api/providers/fetchers/modelCache.ts
   - packages/types/src/providers/index.ts
   - webview-ui/src/components/settings/providers/index.ts
   - webview-ui/src/components/settings/constants.ts
   - webview-ui/src/i18n/locales/en/settings.json

2. Added comprehensive test suite:
   - src/api/providers/__tests__/aistupidlevel.spec.ts
   - Tests provider initialization and configuration
   - Tests API key validation
   - Tests model selection and routing strategies
   - Tests streaming and single completion modes
   - Tests error handling
   - Tests all routing strategy models (auto, auto-coding, auto-reasoning, etc.)
   - Follows same pattern as synthetic.spec.ts

All changes now properly marked for easier merging from Roo/Cline.

StudioPlatforms committed:
fix: Address maintainer code review feedback

Addresses all inline code review comments:

1. Documentation (apps/kilocode-docs/docs/providers/aistupidlevel.md):
   - Removed marketing copy, replaced with link to aistupidlevel.info
   - Moved 'Provider Keys Required' note to Configuration section
   - Simplified Key Features section with link to website

2. Pricing (packages/types/src/providers/aistupidlevel.ts):
   - Set realistic default prices: $0.50 input, $1.50 output per million tokens
   - Added comment explaining prices are approximate averages
   - Added aiStupidLevelFallbackModels array for default models

3. API Key (src/api/providers/fetchers/aistupidlevel.ts):
   - Made apiKey a required parameter (removed optional)
   - Moved hardcoded fallback models to types file
   - Now uses aiStupidLevelFallbackModels from @roo-code/types
   - Updated default pricing fallbacks to match types

4. Cost Tracking (src/api/providers/aistupidlevel.ts):
   - Removed totalCost: 0 override
   - Added comment explaining cost calculation is handled by pricing
   - Allows proper cost calculation from inputPrice/outputPrice

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

mcowger reviewed (commented):
It looks good to me, but I am not a member of the kilo team.

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

StudioPlatforms committed:
Merge branch 'main' into add-aistupidlevel-provider

chrarnoldus commented:
Closing because we don't seen much demand for this provider in our community. We may revisit if that changes.
